### PR TITLE
cephadm: add prometheus adopt

### PR DESCRIPTION
**Pull request by swiftgist**
Follow the same strategy for migrating an existing prometheus installation into the container structure. Keep logic separate to avoid mixing Ceph and non-Ceph behavior. Refactor duplication in later PRs.

Signed-off-by: Eric Jackson <swiftgist@gmail.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by liewegas** [comment]
i've fixed the corpus commits.. can you adjust this PR to reference the latest master there?

**Comment by swiftgist** [comment]
> i've fixed the corpus commits.. can you adjust this PR to reference the latest master there?

I think that did it.

File: src/cephadm/cephadm [hunk 1]
```diff
@@ -2004,6 +2043,20 @@ def command_bootstrap():
...
+    if daemon_type == 'prometheus':
+        uid, gid = extract_uid_gid(file_path='/etc/prometheus')
+    elif daemon_type == 'node-exporter':
+        uid, gid = 65534, 65534
+    elif daemon_type == 'grafana':
+        uid, gid = extract_uid_gid(file_path='/var/lib/grafana')
+    else:
+        raise Error("{} not implemented yet".format(daemon_type))
+    return uid, gid
+
```
Comment (sebastian-philipp) [inline src/cephadm/cephadm]: Maybe we can make the other way around? What about calling `extract_uid_gid_monitoring` from `extract_uid_gid`?

File: src/cephadm/cephadm [hunk 2]
```diff
@@ -2370,6 +2401,12 @@ def command_adopt():
         call_throws([container_path, 'pull', args.image])
 
     (daemon_type, daemon_id) = args.name.split('.', 1)
+
+    if daemon_type == 'prometheus':
+        (uid, gid) = extract_uid_gid_monitoring(daemon_type)
```
Comment (sebastian-philipp) [inline src/cephadm/cephadm]: If we could call `extract_uid_gid` here instead, would it be possible to just continue with the flow, till after the `l.acquire()` call?

File: src/cephadm/cephadm [hunk 3]
```diff
@@ -2499,6 +2536,54 @@ def command_adopt():
     else:
         raise Error('adoption of style %s not implemented' % args.style)
 
+def command_adopt_prometheus(daemon_id, uid, gid):
```
Comment (sebastian-philipp) [inline src/cephadm/cephadm]: Just thinking out loud, can we make this a `@staticmethod` of `Monitoring`?

**Comment by sebastian-philipp** [comment]
```
Fetched in submodule path 'cephadm-adoption-corpus', but it did not contain 50c5dd734638939facd1ed32295ce59c9a5986b4. Direct fetching of that commit failed.
Build step 'Execute shell' marked build as failure
```

@swiftgist can you simply remove the commit to the submodule? Looks like Sage's force push didn't worked.

**Comment by liewegas** [comment]
I think this commit is fine, actually.  It's just that the test script clones from master and ignores the submodule commit.  (We could/should probably just remove the submodule entirely?)  I'll move the master branch on the other repo.

**Comment by liewegas** [comment]
Updated

**Comment by liewegas** [comment]
```
cephadm:2587: error: Too many values to unpack (2 expected, 3 provided)
cephadm:2589: error: Cannot determine type of 'state'
cephadm:2592: error: Cannot determine type of 'enabled'
```

**Comment by liewegas** [comment]
see also #33587

File: src/cephadm/cephadm (hunk 1 above)
Comment (mgfritch) [reply src/cephadm/cephadm]: and/or lift this into the `Monitoring` class .. ?

File: src/cephadm/cephadm (hunk 3 above)
Comment (mgfritch) [reply src/cephadm/cephadm]: the script is getting rather large and it would be nice to start organizing some of the functionality into classes :+1:

**Pull request merged**